aeneas/src/ir/VariantSolver.v3

```diff
@@ -83,6 +101,197 @@ def ON_STACK = -1;
+	// We can only determine distinguishability after all the scalars have been assigned. Heuristics will help speed this part up.
+	private def tryRepresentationForField(curCase: int, curField: int) -> bool {
+		if (curCase >= normFields.length) {
+			// XXX: check distinguishable
+			return solvePacking();
+		}
+		if (curField >= normFields[curCase].length) {
+			// move on to next case
+			curUsed = Vector<byte>.new();
+			for (i < curRep.length) curUsed.put(0);
+			return tryRepresentationForField(curCase + 1, 0);
+		}
+		
+		var sc = vnorm.nc.GetScalar(vnorm.rn.ra.compiler, vnorm.rn.ra.prog, normFields[curCase][curField]);
+		var bw = vnorm.nc.GetBitWidth(vnorm.rn.ra.compiler, vnorm.rn.ra.prog, normFields[curCase][curField]);
+		var none: Scalar.set;
+
+		for (i < curRep.length) {
+			if (usePacking && curUsed[i] + bw > 64) continue;
+			if (!usePacking && curUsed[i] > 0) continue;
+			var oldRep = curRep[i];
+
+			var newRep = sc & oldRep;
+			for (j in sc & oldRep) if (usePacking && j.size < curUsed[i] + bw) newRep -= j;
+
+			if (newRep != none) {
+				curUsed[i] += bw; curRep[i] = newRep;
+				scalarAsns[CaseField(curCase, curField)] = i;
+				if (tryRepresentationForField(curCase, curField + 1)) return true;
+				curUsed[i] -= bw; curRep[i] = oldRep;
+			}
+		}
+
+		var len = curRep.length;
+		curRep.put(sc); curUsed.put(bw);
+		scalarAsns[CaseField(curCase, curField)] = curRep.length - 1;
+		if (tryRepresentationForField(curCase, curField + 1)) return true;
+		curRep.resize(len); curUsed.resize(len);
+
+		return false;
+	}
+	private def tryExplicitTaggingHeuristic() -> bool {
+		for (i < state[0].scalars.length) {
+			var scalar0 = state[0].scalars[i];
+			var longest: Interval, curStart: byte = 0;
+			for (j < scalar0.bits.length) {
+				var allUnassigned = true;
+				for (k < state.length) {
+					if (state[k].scalars[i].bits[j] != PackingBit.Unassigned) {
+						allUnassigned = false;
+						break;
+					}
+				}
+
+				if (allUnassigned) {
+					if (j + 1 - curStart > longest.size()) longest = Interval(curStart, byte.!(j + 1));
+				} else {
+					curStart = byte.!(j + 1);
+				}
+			}
+
+			if (explicitTagLength > 0 && longest.size() >= explicitTagLength) {
+				for (j < state.length) {
+					for (k < explicitTagLength) {
+						var bit = u1.!((j >> k) & 1);
+						state[j].scalars[i].bits[longest.start + k] = PackingBit.Fixed(bit);
+					}
+				}
+				explicitTag = (i, Interval(longest.start, longest.start + explicitTagLength));
+				return true;
+			}
+		}
+		return false;
+	}
+	private def canDistinguish(active: Array<bool>) {
+		// TODO - this is where unassigned bits will get assigned
+	}
+	private def checkDistinguishable() -> bool {
+		if (useTagScalar) return true;
+		if (normFields.length <= 1) return true; // single case is always distinguishable
+		var result = tryExplicitTaggingHeuristic();
+		// TODO: try difficult tagging
+		return result;
+	}
+	private def solveField(idx: int) -> bool {
```

Comment:
I'm starting to wonder if this would be cleaner to just do an iterative solution, e.g. by sorting and using first-first in decreasing order. Also, we're repeatedly getting the width of a field; seems like that should be cached internally in the solver.

Reply:
What do you mean here by first-first in decreasing order? I can try the iterative method, but it would be tricky to do the recursive backtracking iteratively (i.e. trying a new interval to slot a field in if the original interval doesn't work).

Reply:
Sorry, I meant "first-fit" in decreasing order. But I guess this is landable in its current form.